prov/sockets/src/sock_progress.c

```diff
@@ -1826,25 +1826,31 @@ static int sock_pe_progress_tx_atomic(struct sock_pe *pe,
 		return 0;
 	len += entry_len;
 
-	/* cmp data */
 	datatype_sz = fi_datatype_size(pe_entry->pe.tx.tx_op.atomic.datatype);
-	for (i = 0; i < pe_entry->pe.tx.tx_op.atomic.cmp_iov_len; i++) {
+	if (pe_entry->flags & FI_INJECT) {
+		/* cmp data */
 		if (sock_pe_send_field(pe_entry,
-				       (void *) (uintptr_t) pe_entry->pe.tx.tx_iov[i].cmp.ioc.addr,
-				       pe_entry->pe.tx.tx_iov[i].cmp.ioc.count *
-				       datatype_sz, len))
+					&pe_entry->pe.tx.inject[0],
+					pe_entry->pe.tx.tx_op.atomic.cmp_iov_len, len))
```

Comment:
@shantonu - For the inject case, you are writing cmp_data at tx.inject[0] + src_len. But, in here, it is read from tx.inject[0].